File: set_of_stacks.cpp
#include <iostream>

using namespace std;

#define NULL_VALUE -1

template <class T>
class Node{
public:
	Node();
	~Node();
	T* data;               
	Node* next;
};

template <class T>
Node<T>::Node(){
	next = NULL;
}
template <class T>
Node<T>::~Node(){
	delete data;
}

template <class T>
class Stack{
public:
	Stack();
	~Stack();
	void push(T*);
	T* pop();
	bool isEmpty();
	Node<T>* peekNode();
	T* peek();
private:
	Node<T>* head;
};

template <class T>
Stack<T>::Stack(){
	head = NULL;
}
template <class T>
Stack<T>::~Stack(){
	while(!isEmpty()){
		Node<T>* del = head;
		head = head->next;
		delete del;
	}
}
template <class T>
void Stack<T>::push(T* data){
	Node<T>* node = new Node<T>;
	node->data = data;
	node->next = head;
	head = node;
}
template <class T>
T* Stack<T>::pop(){				
	if(isEmpty()){    
		return NULL;
	}
	T* data;
	data = head->data;
	head = head->next;
	return data;
}
template <class T>
bool Stack<T>::isEmpty(){
	return head == NULL;
}
template <class T>
Node<T>* Stack<T>::peekNode(){
	return head;
}
template <class T>
T* Stack<T>::peek(){
	return Stack::isEmpty() ? NULL : head->data;
}

class SetOfStacks {
public:
	SetOfStacks();
	~SetOfStacks();
	void push(int);
	int pop();
	int peek();
	bool isEmpty();
	int popAt(int);
private:
	Stack<Stack<int>>* _setOfStacks;
	bool isThereChildStack();
	int counter;
	int threshold;
	int subIndex;
};

SetOfStacks::SetOfStacks(){	
	_setOfStacks = new Stack<Stack<int>>();
	counter = 0;
	threshold = 3;
	subIndex = -1;	
}

SetOfStacks::~SetOfStacks(){
	delete _setOfStacks;
}

void SetOfStacks::push(int data){
	if(_setOfStacks->isEmpty() || counter == threshold){
		Stack<int>* newStack = new Stack<int>();
		_setOfStacks->push(newStack);
		subIndex = subIndex + 1;
		counter = 0;
	}
	_setOfStacks->peek()->push(new int {data});
	counter = counter + 1;
}

int SetOfStacks::pop(){
	if(!isThereChildStack()){
		return NULL_VALUE;
	}
    int* data = _setOfStacks->peek()->pop();
	counter = counter - 1;
    if(data == NULL){
    	return -1;
    }
	return *data;
}

int SetOfStacks::peek(){
	if(!isThereChildStack()){
		return NULL_VALUE;
	}
	return _setOfStacks->isEmpty() ? NULL_VALUE : *_setOfStacks->peek()->peek();
}

bool SetOfStacks::isEmpty(){
	return _setOfStacks->isEmpty();
}

bool SetOfStacks::isThereChildStack(){
	if(counter==0){
		_setOfStacks->pop();
		if(_setOfStacks->isEmpty()){
			subIndex = -1;
			return false;
		}
		subIndex = subIndex - 1;
		counter = threshold;
	}
	return true;	
}

int SetOfStacks::popAt(int subStack){
	int diff = subIndex - subStack;
	if(diff < 0){
		return NULL_VALUE;
	}
	Stack<Stack<int>>* tempStack = new Stack<Stack<int>>();
	for(int i=0; i < diff; i++){
		tempStack->push(_setOfStacks->pop());
	}
	int data = *_setOfStacks->peek()->pop(); 
	for(int i=0; i < diff; i++){
		_setOfStacks->push(tempStack->pop());
	}
	delete tempStack; 
	return data;
}

int main(){

	SetOfStacks* stack = new SetOfStacks();
	cout << "isEmpty" << stack->isEmpty() << endl;
	stack->push(6);
	stack->push(1);
	stack->push(4);
	stack->push(9);
	stack->push(8);
	stack->push(888);
	stack->push(49);
	stack->push(94);
	stack->push(56);
	stack->push(87);
	cout << "isEmpty " << stack->isEmpty() << endl;
	cout << stack->peek() << endl;
	cout << "popAt " << stack->popAt(0) << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << "popAt " << stack->popAt(5) << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << "peek " << stack->peek() << endl;
	cout << stack->pop() << endl;
	cout << "popAt " << stack->popAt(1) << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << "peek " << stack->peek() << endl;
	cout << stack->pop() << endl;
	cout << stack->pop() << endl;
	cout << "peek " << stack->peek() << endl;
	cout << stack->pop() << endl;

	return 0;
}

File: set_of_stacks.py
from stack import Stack

class SetOfStacks(object):
	def __init__(self):
		self.setof_stacks = Stack()
		self.counter = 0
		self.threshold = 3
		self.sub_index = 0

	def _is_there_child_stack(self):
		if self.counter == 0:
			self.setof_stacks.pop()
			if self.setof_stacks.is_empty():
				return False
			self.sub_index = self.sub_index - 1
			self.counter = self.threshold
		return True

	def push(self, data):
		if self.setof_stacks.is_empty() or self.counter == self.threshold:
			self.setof_stacks.push(Stack())
			self.sub_index = self.sub_index + 1
			self.counter = 0 					
		self.setof_stacks.peek().push(data)
		self.counter = self.counter + 1

	def pop(self):
		if not self._is_there_child_stack():
			return None
		data = self.setof_stacks.peek().pop()
		self.counter = self.counter - 1
		return data 

	def peek(self):
		if not self._is_there_child_stack():
			return None
		return self.setof_stacks.peek().peek()

	def is_empty(self):
		return self.setof_stacks.is_empty()
	
	def pop_at(self, sub_stack):
		diff = self.sub_index - sub_stack
		if diff < 0:
			return None
		temp_stack = Stack()
		for _ in xrange(diff):
			temp_stack.push(self.setof_stacks.pop())
		dt = self.setof_stacks.peek().pop()
		for _ in xrange(diff):
			self.setof_stacks.push(temp_stack.pop())
		return dt

def main():

	stack = SetOfStacks()
	# print(stack.pop())
	# print(stack.peek())	
	# print(stack.is_empty())
	stack.push(3)
	print("counter", stack.counter)
	# stack.push(5)
	# print("counter", stack.counter)
	# stack.push(3)
	# print("counter", stack.counter)
	# stack.push(2)
	# print("counter", stack.counter)
	# print(stack.pop())
	# print(stack.peek())
	# stack.push(9)
	# print("counter", stack.counter)
	# stack.push(5)
	# print("counter", stack.counter)
	# print(stack.pop())
	# print(stack.pop())
	print("popat", stack.pop_at(1))
	# print("popat", stack.pop_at(1))
	# print("popat", stack.pop_at(1))
	# print("popat", stack.pop_at(1))
	print(stack.pop())
	# print(stack.is_empty())
	# # print(stack.pop())
	# print(stack.is_empty())
	# # print(stack.pop())
	# # print(stack.pop())
	# stack.push(6)
	# stack.push(7)
	# print("popat", stack.pop_at(2))
	# print("popat", stack.pop_at(2))
	# print("popat", stack.pop_at(3))
	# print("popat", stack.pop_at(2))
	# print("self", stack.sub_index)
	# print("popat", stack.pop_at(5))
	# print(stack.peek())
	# print(stack.pop())
	# # print(stack.peek())
	# # print(stack.pop())
	# stack.push(6)
	# stack.push(7)
	# stack.push(6)
	# stack.push(7)
	# print("popat", stack.pop_at(3))
	# # print(stack.pop())
	# print(stack.pop())
	# print(stack.pop())
	# print(stack.pop())
	# print(stack.is_empty())
	# print(stack.pop())
	# print(stack.pop())
	# print(stack.is_empty())
	# print(stack.peek())
	# stack.push(5)
	# stack.push(1)
	# stack.push(2)
	# stack.push(9)
	# print(stack.pop())
	# print(stack.pop())
	# print(stack.pop())
	# print(stack.pop())


if __name__ == "__main__":
	main()
